rotkehlchen/chain/evm/decoding/open_ocean/decoder.py

```diff
@@ -0,0 +1,188 @@
+import logging
+from abc import ABC
+from collections.abc import Callable
+from typing import TYPE_CHECKING, Any, Final
+
+from rotkehlchen.chain.ethereum.utils import (
+    token_normalized_value,
+    token_normalized_value_decimals,
+)
+from rotkehlchen.chain.evm.decoding.interfaces import DecoderInterface
+from rotkehlchen.chain.evm.decoding.structures import (
+    DEFAULT_DECODING_OUTPUT,
+    DecoderContext,
+    DecodingOutput,
+)
+from rotkehlchen.chain.evm.decoding.types import CounterpartyDetails
+from rotkehlchen.chain.evm.decoding.uniswap.constants import (
+    UNISWAP_V2_SWAP_SIGNATURE,
+    UNISWAP_V3_SWAP_SIGNATURE,
+)
+from rotkehlchen.chain.evm.decoding.utils import maybe_reshuffle_events
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChecksumEvmAddress, EvmTransaction
+from rotkehlchen.utils.misc import bytes_to_address
+
+from .constants import CPT_OPENOCEAN, OPENOCEAN_EXCHANGE_ADDRESS, OPENOCEAN_LABEL
+
+if TYPE_CHECKING:
+    from rotkehlchen.assets.asset import Asset, CryptoAsset
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.chain.evm.structures import EvmTxReceiptLog
+    from rotkehlchen.fval import FVal
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+    from rotkehlchen.user_messages import MessagesAggregator
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+SWAPPED_TOPIC: Final = b'v\xaf"J\x148e\xa5\x0bAIn\x1asb&\x98i,V\\\x12\x14\xbc\x86/\x18\xe2-\x82\x9c^'  # noqa: E501
+
+
+class OpenOceanCommonDecoder(DecoderInterface, ABC):
+
```

Comment:
I read that there is difference between this and `metaclass=ABC`. Maybe worth reading since I don't have all the details

Reply:
As far as I can tell simply using `ABC` is correct. That is the first way its shown here: https://docs.python.org/3/library/abc.html, although it does also show `metaclass=ABCMeta` as being valid as well.